test/state/precompiles.cpp

```diff
@@ -186,34 +168,75 @@ PrecompileAnalysis bls12_g1mul_analyze(bytes_view, evmc_revision) noexcept
     return {BLS12_G1MUL_PRECOMPILE_GAS, 128};
 }
 
+static consteval auto bls12_msm_discounts_to_point_costs(
+    int mul_gas_cost, std::span<const int, 128> discounts) noexcept
+{
+    std::array<int16_t, std::size(discounts)> cost{};
+    for (size_t i = 0; i < cost.size(); ++i)
+    {
+        const auto c = mul_gas_cost * discounts[i] / 1000;
+        assert(c <= std::numeric_limits<int16_t>::max());
+        cost[i] = static_cast<int16_t>(c);
+    }
+    return cost;
+}
+
 PrecompileAnalysis bls12_g1msm_analyze(bytes_view input, evmc_revision) noexcept
 {
+    static constexpr auto G1MUL_GAS_COST = 12000;
+    static constexpr int DISCOUNTS[] = {1000, 949, 848, 797, 764, 750, 738, 728, 719, 712, 705, 698,
+        692, 687, 682, 677, 673, 669, 665, 661, 658, 654, 651, 648, 645, 642, 640, 637, 635, 632,
+        630, 627, 625, 623, 621, 619, 617, 615, 613, 611, 609, 608, 606, 604, 603, 601, 599, 598,
+        596, 595, 593, 592, 591, 589, 588, 586, 585, 584, 582, 581, 580, 579, 577, 576, 575, 574,
+        573, 572, 570, 569, 568, 567, 566, 565, 564, 563, 562, 561, 560, 559, 558, 557, 556, 555,
+        554, 553, 552, 551, 550, 549, 548, 547, 547, 546, 545, 544, 543, 542, 541, 540, 540, 539,
+        538, 537, 536, 536, 535, 534, 533, 532, 532, 531, 530, 529, 528, 528, 527, 526, 525, 525,
+        524, 523, 522, 522, 521, 520, 520, 519};
+    static constexpr auto POINT_COSTS =
+        bls12_msm_discounts_to_point_costs(G1MUL_GAS_COST, DISCOUNTS);
+
     if (input.empty() || input.size() % 160 != 0)
```

Comment:
Can we define single input size and use ut here and below? Same for G2

Reply:
I defined constants for sizes, but not used it `&input[64]`... I think the input handling may need separate improvement.